### PR TITLE
Return device count using nvml

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84879
* #84878
* #84877

Fixes https://github.com/pytorch/pytorch/issues/83973

### CODE REVIEW

File: torch/cuda/__init__.py [hunk 1]
```diff
@@ -461,12 +461,62 @@ def set_stream(stream: Stream):
...
+            return -1
+        for idx, c in enumerate(s):
+            if not c.isdigit():
+                break
+            if idx + 1 == len(s):
+                idx += 1
+        return int(s[:idx]) if idx > 0 else -1
+
+    # CUDA_VISIBLE_DEVICES uses something like strtoul
+    # which makes `1gpu2,2ampere` is equivalent to `1,2`
```
Comment (ngimel) [inline torch/cuda/__init__.py:481]: til

Comment (ngimel) [reply torch/cuda/__init__.py:481]: This doesn't handle things like 0-2?

**Review by ngimel — APPROVED** [review]

File: torch/cuda/__init__.py (hunk 1 above)
Comment (malfet) [reply torch/cuda/__init__.py:481]: Nope, at using 510 driver, ranges do not seem to be supported (prior to this change):
```

**Pull request closed without merging**